File: 0039-combination-sum/0039-combination-sum.java
public class Solution {
public List<List<Integer>> combinationSum(int[] candidates, int target) {
    List<List<Integer>>[] dp = new List[target + 1];
    for (int i = 0; i <= target; i++)
        dp[i] = new ArrayList<>();
    dp[0].add(new ArrayList<>());
    for (int candidate: candidates) {
        for (int j = candidate; j <= target; j++) {                
            for (List<Integer> comb: dp[j - candidate]) {
                List<Integer> newComb = new ArrayList(comb);
                newComb.add(candidate);
                dp[j].add(newComb);
            }                    
        }
    }
    return dp[target];
}
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        results = []

        def bt(remain, comb, start):
            if remain < 0:
                return
            if remain == 0:
                results.append(list(comb))
                return

            for i in range(start, len(candidates)):
                n = candidates[i]
                comb.append(n)
                bt(remain-n, comb, i)
                comb.pop()

        bt(target, [], 0)
        return results